core/src/database/clickhouse/client.rs

```diff
@@ -0,0 +1,115 @@
+use crate::EthereumSqlTypeWrapper;
+use clickhouse::{Client, Row};
+use dotenv::dotenv;
+use serde::Deserialize;
+use std::env;
+use tracing::info;
+
+pub struct ClickhouseConnection {
+    url: String,
+    user: String,
+    password: String,
+    db: String,
+}
+
+pub fn clickhouse_connection() -> Result<ClickhouseConnection, env::VarError> {
+    dotenv().ok();
+
+    let connection = ClickhouseConnection {
+        url: env::var("CLICKHOUSE_URL")?,
+        user: env::var("CLICKHOUSE_USER")?,
+        password: env::var("CLICKHOUSE_PASSWORD")?,
+        db: env::var("CLICKHOUSE_DB")?,
+    };
+
+    Ok(connection)
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum ClickhouseConnectionError {
+    #[error("The clickhouse env vars are wrong please check your environment: {0}")]
+    ClickhouseConnectionConfigWrong(#[from] env::VarError),
+
+    #[error("Could not connect to clickhouse database: {0}")]
+    ClickhouseNetworkError(#[from] clickhouse::error::Error),
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum ClickhouseError {
+    #[error("ClickhouseError {0}")]
+    ClickhouseError(String),
+}
+
+pub struct ClickhouseClient {
+    conn: Client,
+}
+
+impl ClickhouseClient {
+    pub async fn new() -> Result<Self, ClickhouseConnectionError> {
+        let connection = clickhouse_connection()?;
+
+        let client = Client::default()
+            .with_url(connection.url)
+            .with_user(connection.user)
+            .with_database(connection.db)
+            .with_password(connection.password);
+
+        client.query("select 1").execute().await?;
+        info!("Clickhouse client connected successfully!");
+
+        Ok(ClickhouseClient { conn: client })
+    }
+
+    pub async fn query_one<T>(&self, sql: &str) -> Result<T, ClickhouseError>
+    where
+        T: Row + for<'b> Deserialize<'b>,
+    {
+        self.conn
+            .query(sql)
+            .fetch_one()
+            .await
+            .map_err(|e| ClickhouseError::ClickhouseError(e.to_string()))
+    }
+
+    pub async fn execute(&self, sql: &str) -> Result<(), ClickhouseError> {
+        self.conn
+            .query(sql)
+            .execute()
+            .await
+            .map_err(|e| ClickhouseError::ClickhouseError(e.to_string()))
+    }
+
+    pub async fn execute_batch(&self, sql: &str) -> Result<(), ClickhouseError> {
+        let statements: Vec<&str> =
+            sql.split(';').map(str::trim).filter(|s| !s.is_empty()).collect();
+
+        for statement in statements {
+            self.execute(statement).await?;
+        }
+
+        Ok(())
+    }
+
+    pub async fn bulk_insert(
```

Comment:
it may be worth a quick trait here with basic required methods at least then we start being able to understand bare bones of a db layer

Reply:
Think I will leave this for now since it seems the implementation for both clients will keep evolving for a bit. Probably makes sense at some point but for now they are different `impls` and technically could have different interfaces if we want to.